ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/ProtoBufDiffy.scala

```diff
@@ -37,28 +38,45 @@ class ProtoBufDiffy[T <: AbstractMessage : ClassTag](ignore: Set[String] = Set.e
       .getMethod("getDescriptor")
       .invoke(null).asInstanceOf[Descriptor]
 
-  // scalastyle:off cyclomatic.complexity
+  // scalastyle:off cyclomatic.complexity method.length
   private def diff(x: AbstractMessage, y: AbstractMessage,
                    fields: Seq[FieldDescriptor], root: String): Seq[Delta] = {
-    def getField(m: AbstractMessage, f: FieldDescriptor): AnyRef =
+    def getField(f: FieldDescriptor)(m: AbstractMessage): AnyRef =
       if (f.isRepeated) {
         m.getField(f)
       } else {
         if (m.hasField(f)) m.getField(f) else null
       }
 
+    def getFieldDescriptor(f: FieldDescriptor, s: String): FieldDescriptor = {
+      f.getMessageType.getFields.asScala.find(field => field.getName == s) match {
+        case Some(d) => d
+        case None => throw new Exception(s"Field $s not found in ${f.getFullName}")
+      }
+    }
+
     fields.flatMap { f =>
       val name = f.getName
       val fullName = if (root.isEmpty) name else root + "." + name
       if (f.isRepeated && unordered.contains(fullName)) {
-        val a = sortList(x.getField(f).asInstanceOf[java.util.List[AnyRef]])
-        val b = sortList(y.getField(f).asInstanceOf[java.util.List[AnyRef]])
-        if (a == b) Nil else Seq(Delta(fullName, a, b, delta(a, b)))
+        val a = sortList(x.getField(f).asInstanceOf[java.util.List[AbstractMessage]],
+          unorderedFieldKeys.get(fullName).map(s => getField(getFieldDescriptor(f, s)) _))
+        val b = sortList(y.getField(f).asInstanceOf[java.util.List[AbstractMessage]],
+          unorderedFieldKeys.get(fullName).map(s => getField(getFieldDescriptor(f, s)) _))
+        if (f.getJavaType == JavaType.MESSAGE && unordered.exists(_.startsWith(s"$fullName."))
+            && unorderedFieldKeys.contains(fullName)) {
+          a.asInstanceOf[java.util.List[AbstractMessage]].asScala.zip(
```

Comment:
do we need to keep converting from Java to Scala and back

Reply:
Yea in this case it's a Java List that we get from protobuf and we need to cast it before wrapping in Scala collection